src/main/kotlin/com/ajaxproject/warehouse/service/WaybillServiceImpl.kt

```diff
@@ -1,130 +1,129 @@
 import com.ajaxproject.warehouse.dto.WaybillDataLiteDto
 import com.ajaxproject.warehouse.dto.WaybillInfoUpdateDto
-import com.ajaxproject.warehouse.entity.Product
-import com.ajaxproject.warehouse.entity.Waybill
-import com.ajaxproject.warehouse.entity.WaybillProduct
+import com.ajaxproject.warehouse.entity.MongoCustomer
+import com.ajaxproject.warehouse.entity.MongoProduct
+import com.ajaxproject.warehouse.entity.MongoWaybill
+import com.ajaxproject.warehouse.exception.InternalEntityNotFoundException
 import com.ajaxproject.warehouse.exception.NotFoundException
-import com.ajaxproject.warehouse.repository.CustomerRepository
-import com.ajaxproject.warehouse.repository.ProductRepository
-import com.ajaxproject.warehouse.repository.WaybillProductRepository
-import com.ajaxproject.warehouse.repository.WaybillRepository
+import com.ajaxproject.warehouse.repository.MongoCustomerRepository
+import com.ajaxproject.warehouse.repository.MongoProductRepository
+import com.ajaxproject.warehouse.repository.MongoWaybillRepository
+import org.bson.types.ObjectId
 import org.springframework.stereotype.Service
 import java.time.LocalDate
 
 @Service
 @Suppress("TooManyFunctions")
 class WaybillServiceImpl(
-    val waybillRepository: WaybillRepository,
-    val customerRepository: CustomerRepository,
-    val waybillProductRepository: WaybillProductRepository,
-    val productRepository: ProductRepository
+    val mongoWaybillRepository: MongoWaybillRepository,
+    val mongoProductRepository: MongoProductRepository,
+    val mongoCustomerRepository: MongoCustomerRepository
 ) : WaybillService {
 
-    override fun findAll(): List<WaybillDataLiteDto> {
-        return waybillRepository.findAll().map { it.mapToLiteDto() }
+    override fun findAllWaybills(): List<WaybillDataLiteDto> =
+        mongoWaybillRepository.findAll().map { it.mapToLiteDto() }
+
+    override fun getById(id: String): WaybillDataDto {
+        val waybill: MongoWaybill = mongoWaybillRepository.findById(ObjectId(id))
+            ?: throw NotFoundException("Waybill with id $id not found")
+        return waybill.mapToDataDto()
     }
 
-    override fun findById(id: Int): WaybillDataDto {
-        return waybillRepository.findById(id)
-            .orElseThrow { NotFoundException("Waybill with id $id not found") }
-            .mapToDataDto()
+    override fun updateWaybillInfo(infoUpdateDto: WaybillInfoUpdateDto, id: String): WaybillDataDto {
+        require(infoUpdateDto.id == id) { "Mapping id is not equal to request body id" }
+        var mongoWaybill: MongoWaybill = mongoWaybillRepository.findById(ObjectId(id))
+            ?: throw NotFoundException("Waybill with id $id not found")
+        mongoCustomerRepository.findById(ObjectId(infoUpdateDto.customerId))
+            ?: throw NotFoundException("Customer with id ${infoUpdateDto.customerId} not found")
+        mongoWaybill = mongoWaybillRepository.save(mongoWaybill.setUpdatedData(infoUpdateDto))
```

Comment:
What if:
1. [Thread-1] customer check passed: `mongoCustomerRepository.findById(ObjectId(infoUpdateDto.customerId))`
2. [Thread-2] Some another thread removed the customer
3. [Thread-1] The updated wayBill is saved

So, it's possible that your DB data could be in inconsistent state

Reply:
Can I use ClientSession.startTransaction() etc. to solve this?

Reply:
And do I need to implement same thread-safe logic in CustomerService.update() and ProductService.update()?